### PR TITLE
Route Sync job

### PR DESCRIPTION
**Pull request by alex-slynko**
there is some code for HTTP route sync but it needs some improvements

### CODE REVIEW

File: jobs/kubernetes-api-route-registrar/templates/bin/announce_route.erb [hunk 1]
```diff
@@ -16,15 +16,15 @@ BACKEND_IP="<%= spec.ip %>"
 KUBO_PORT="<%=p('external_kubo_port') %>"
 
 # Fetch UAA token
-UAA_REQUEST=$(curl -s -k -X POST ${UAA_API_URL}/oauth/token --data "client_id=${UAA_CLIENT_ID}&client_secret=${UAA_CLIENT_SECRET}&grant_type=client_credentials&response_type=token")
+UAA_REQUEST=$(curl --fail -s -k -X POST ${UAA_API_URL}/oauth/token --data "client_id=${UAA_CLIENT_ID}&client_secret=${UAA_CLIENT_SECRET}&grant_type=client_credentials&response_type=token")
```
Comment (alex-slynko) [inline jobs/kubernetes-api-route-registrar/templates/bin/announce_route.erb]: Extracted this to a separate PR #25

File: jobs/route-sync/spec [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      example: http://api.cf.cf-app.com
+    uaa_url:
+      description: "API URL for UAA"
+      example: http://uaa.cf.cf-app.com
+    routing_api_client_id:
+      description: "Routing API client id (UAA)"
+      example: tcp_router
+    routing_api_client_secret:
+      description: "Routing API client secret"
+      example: "kljsfdkjldfskjlfsdsfd"
```
Comment (alex-slynko) [inline jobs/route-sync/spec]: Example is not required. Especially this one

File: jobs/route-sync/spec [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    skip_tls_verification:
```
Comment (alex-slynko) [inline jobs/route-sync/spec]: We need default value for this

File: jobs/route-sync/spec [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    description: IPs of each NATS cluster member
+    example: |
+      - 192.168.50.123
+      - 192.168.52.123
+  nats.port:
+    description: TCP port of NATS servers
+    example: 4222
+  nats.user:
+    description: User name for NATS authentication
+    example: nats
```
Comment (alex-slynko) [inline jobs/route-sync/spec]: default value, not example here as well

File: jobs/route-sync/spec [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      example: app.cf.cf-app.com
+ 
+  nats.machines:
+    description: IPs of each NATS cluster member
+    example: |
+      - 192.168.50.123
+      - 192.168.52.123
+  nats.port:
+    description: TCP port of NATS servers
+    example: 4222
```
Comment (alex-slynko) [inline jobs/route-sync/spec]: default value, not example here

File: jobs/route-sync/templates/bin/route-sync-ctl.erb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+setup_directories() {
+  mkdir -p "$RUN_DIR" "$LOG_DIR" "$DATA_DIR"
+  chown -R vcap:vcap "$RUN_DIR" "$LOG_DIR" "$DATA_DIR"
+}
+
+start_registrar() {
+  pid_guard $PIDFILE route-sync
+  echo $$ > $PIDFILE
+
+  export ROUTESYNC_CLOUD_FOUNDRY_API_URL="<%=p("cloud_foundry.api_url") %>"
```
Comment (alex-slynko) [inline jobs/route-sync/templates/bin/route-sync-ctl.erb]: Little bit curious why environment variables and not config/flags

File: jobs/route-sync/templates/bin/route-sync-ctl.erb [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+pid() {
+  head -1 "$PIDFILE"
+}
+
+case $1 in
+
+  start)
+    setup_directories
+    start_registrar
```
Comment (alex-slynko) [inline jobs/route-sync/templates/bin/route-sync-ctl.erb]: it is not registrar

File: src/route-sync/README.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
+# route-sync
+
+Synchronizes routes from a Kubernetes cluster to external L3/L7 routers.
+
+## Dependencies
+
+- golang 1.7.4
```
Comment (alex-slynko) [inline src/route-sync/README.md]: 1.8

File: src/route-sync/README.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+Synchronizes routes from a Kubernetes cluster to external L3/L7 routers.
+
+## Dependencies
+
+- golang 1.7.4 
+
+## Development
+
+This repo should be imported as `route-sync` meaning your GOPATH should be set to `~/kubo-release`
```
Comment (alex-slynko) [inline src/route-sync/README.md]: it is not always ~. Better root directory for `kubo-release`

File: src/route-sync/cloudfoundry/router.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+		return tcp.RouterGroup{}, fmt.Errorf("NYI: Multiple router groups not supported")
+	}
+
+	return routerGroups[0], nil
+}
+
+func (ts *router) HTTP(routes []*route.HTTP) error {
+	for _, httpRoute := range routes {
+		for _, endpoint := range httpRoute.Backend {
+			err := ts.bus.SendMessage("router.register", endpoint.IP, config.Route{
```
Comment (alex-slynko) [inline src/route-sync/cloudfoundry/router.go]: It would be nice to have same (or almost same) interface for registering TCP & HTTP routes.
SendMessage requires knowledge of GoRouter in this code.

File: src/route-sync/cloudfoundry/router.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+func (ts *router) tcpRouterGroup() (tcp.RouterGroup, error) {
+	routerGroups, err := ts.tcpRouter.RouterGroups()
+
+	if err != nil {
+		return tcp.RouterGroup{}, err
+	}
+
+	if len(routerGroups) != 1 {
```
Comment (alex-slynko) [inline src/route-sync/cloudfoundry/router.go]: don't see tests for this

File: src/route-sync/cloudfoundry/router.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+		return err
+	}
+
+	return ts.tcpRouter.CreateRoutes(routerGroup, routes)
+}
+
+func (ts *router) tcpRouterGroup() (tcp.RouterGroup, error) {
+	routerGroups, err := ts.tcpRouter.RouterGroups()
+
+	if err != nil {
```
Comment (alex-slynko) [inline src/route-sync/cloudfoundry/router.go]: tests :(

File: src/route-sync/cloudfoundry/tcp/router.go [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+type Router interface {
+	RouterGroups() ([]RouterGroup, error)
+	CreateRoutes(RouterGroup, []*route.TCP) error
+}
+
+const routeTTL = 60
+
+type routing_api_router struct {
+	uaaClient           uaa_go_client.Client
+	cloudFoundryApiUrl  string
```
Comment (alex-slynko) [inline src/route-sync/cloudfoundry/tcp/router.go]: it is routing api url (which usually is the same as cloudfoundry api url)

File: src/route-sync/cloudfoundry/tcp/router_test.go [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+		var invalidCreations = []struct {
+			uaaClient uaa_go_client.Client
+			api       string
+		}{
+			{nil, ""},
+			{&fooClient, ""},
+			{nil, "bar"},
+		}
+
+		for _, t := range invalidCreations {
```
Comment (alex-slynko) [inline src/route-sync/cloudfoundry/tcp/router_test.go]: I like this 🥇 
Do this create multiple tests?

File: src/route-sync/cloudfoundry/tcp/router_test.go [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+			Context(fmt.Sprintf("uaaClient: %q, api: %s", t.uaaClient, t.api), func() {
+				router, err := NewRoutingApi(t.uaaClient, t.api, false)
+				Expect(router).To(BeNil())
+				Expect(err).NotTo(BeNil())
+			})
+		}
+
+		router, err := NewRoutingApi(&fooClient, "foo", false)
```
Comment (alex-slynko) [inline src/route-sync/cloudfoundry/tcp/router_test.go]: this feels like different positive test which has to be separated from negative cases

File: src/route-sync/cloudfoundry/tcp/router_test.go [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+		}))
+		defer ts.Close()
+
+		router, _ := NewRoutingApi(&fooClient, ts.URL, false)
+
+		router.RouterGroups()
+		req := <-requestChan
+		Expect(req.Header).To(HaveKeyWithValue("Authorization", []string{"bearer foo"}))
+
+		router.CreateRoutes(RouterGroup{}, []*route.TCP{&route.TCP{}})
```
Comment (alex-slynko) [inline src/route-sync/cloudfoundry/tcp/router_test.go]: differen test case please

File: src/route-sync/cloudfoundry/tcp/router_test.go [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+		req = <-requestChan
+		Expect(req.Header).To(HaveKeyWithValue("Authorization", []string{"bearer foo"}))
+
+	})
+	It("handles routing groups", func() {
```
Comment (alex-slynko) [inline src/route-sync/cloudfoundry/tcp/router_test.go]: maybe 
```
Context("RouterGroups"
```
You can add header test here

File: src/route-sync/cloudfoundry/tcp/router.go [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+	req.Header.Set("Authorization", fmt.Sprintf("bearer %s", key))
+
+	return req, client, nil
+}
+
+func (r *routing_api_router) RouterGroups() ([]RouterGroup, error) {
+	var routerGroups []RouterGroup
+
+	req, client, err := r.buildRequest("GET", "routing/v1/router_groups")
+	if err != nil {
```
Comment (alex-slynko) [inline src/route-sync/cloudfoundry/tcp/router.go]: errors are not tested :(

File: src/route-sync/cloudfoundry/tcp/router.go [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+		err = fmt.Errorf("routing_api_router: unmarshalling response: %v\n body: %s", err, body)
+	}
+
+	return routerGroups, err
+}
+
+func (r *routing_api_router) CreateRoutes(rg RouterGroup, routes []*route.TCP) error {
+	req, client, err := r.buildRequest("POST", "routing/v1/tcp_routes/create")
+	if err != nil {
```
Comment (alex-slynko) [inline src/route-sync/cloudfoundry/tcp/router.go]: errors 😿

**Comment by jhvhs** [comment]
I would suggest managing dependencies consistently. We also use glide for dependency management, but let's stick to checking in all of the dependencies into the vendor directory, and refrain from adding any submodules.

File: src/route-sync/pooler/pooler_test.go [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+		src.TCP_value = []*route.TCP{tcpRoute}
+		src.HTTP_value = []*route.HTTP{httpRoute}
+		router := &routefakes.Router{}
+
+		done := pool.Start(src, router)
+
+		Eventually(func() bool {
+			src.Lock()
+			defer src.Unlock()
+			return src.TCP_count > 0
```
Comment (jhvhs) [inline src/route-sync/pooler/pooler_test.go]: Would be even better to use counterfeiter for generating the fakes

File: src/route-sync/route/fakes/source.go [hunk 21]
```diff
@@ -0,0 +1,31 @@
+package fakes
```
Comment (jhvhs) [inline src/route-sync/route/fakes/source.go]: See comment above in `pooler_test`

File: src/route-sync/cloudfoundry/router.go (hunk 12 above)
Comment (johnsonj) [inline src/route-sync/cloudfoundry/router.go]: I don't think these checks warrant it

File: src/route-sync/cloudfoundry/tcp/router.go (hunk 13 above)
Comment (johnsonj) [inline src/route-sync/cloudfoundry/tcp/router.go]: Good call, changing throughout.

File: src/route-sync/cloudfoundry/tcp/router_test.go (hunk 14 above)
Comment (johnsonj) [inline src/route-sync/cloudfoundry/tcp/router_test.go]: Thanks! each one should show up as a separate test case.

File: src/route-sync/cloudfoundry/tcp/router_test.go (hunk 15 above)
Comment (johnsonj) [inline src/route-sync/cloudfoundry/tcp/router_test.go]: That's true, in fact I really don't see the value of this positive test. It just performs creation.

File: src/route-sync/cloudfoundry/tcp/router_test.go (hunk 16 above)
Comment (johnsonj) [inline src/route-sync/cloudfoundry/tcp/router_test.go]: :+1:

File: src/route-sync/cloudfoundry/tcp/router_test.go (hunk 17 above)
Comment (johnsonj) [inline src/route-sync/cloudfoundry/tcp/router_test.go]: :+1:

File: packages/golang/packaging [hunk 22]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+set -e # exit immediately if a simple command exits with a non-zero status
+set -u # report the usage of uninitialized variables
+
+tar xzvf ${BOSH_COMPILE_TARGET}/go/go1.8.linux-amd64.tar.gz
+
+mkdir -p ${BOSH_INSTALL_TARGET}/bin
```
Comment (johnsonj) [inline packages/golang/packaging]: It isn't needed. Copy & pasted this from a bosh release that hat it.

File: src/route-sync/cloudfoundry/tcp/router.go [hunk 23]
```diff
@@ -0,0 +1,139 @@
+package tcp
```
Comment (johnsonj) [inline src/route-sync/cloudfoundry/tcp/router.go]: I wasn't aware of it when I did this implementation

File: src/route-sync/cloudfoundry/router.go [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+	}
+
+	if len(routerGroups) != 1 {
+		return tcp.RouterGroup{}, fmt.Errorf("NYI: Multiple router groups not supported")
+	}
+
+	return routerGroups[0], nil
+}
+
+func (ts *router) HTTP(routes []*route.HTTP) error {
```
Comment (johnsonj) [inline src/route-sync/cloudfoundry/router.go]: I went with that for interface consistency. It seems like a method that ought to allow for failure and the caller should be prepared to handle that.

File: src/route-sync/config/config.go [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+	cfConfig "code.cloudfoundry.org/route-registrar/config"
+	"github.com/kelseyhightower/envconfig"
+)
+
+// Config contains settings for the route-sync with envconfig annotations
+type Config struct {
+	RawNatsServers            string                      `envconfig:"nats_servers" required:"true"`
+	NatsServers               []cfConfig.MessageBusServer `ignore:"true"`
+	CloudFoundryAPIURL        string                      `envconfig:"cloud_foundry_api_url" required:"true"`
+	CloudFoundryAPPDomainName string                      `envconfig:"cloud_foundry_app_domain_name" required:"true"`
```
Comment (johnsonj) [inline src/route-sync/config/config.go]: Changed to App

File: src/route-sync/pooler/pooler_test.go (hunk 20 above)
Comment (johnsonj) [inline src/route-sync/pooler/pooler_test.go]: I considered using something to do mocking, but the complexities of setting that up and generating code seemed to be out weight by stubbing these simple interfaces.

File: src/route-sync/cloudfoundry/router.go (hunk 12 above)
Comment (johnsonj) [inline src/route-sync/cloudfoundry/router.go]: discussed it, it does i think

File: src/route-sync/cloudfoundry/router.go (hunk 11 above)
Comment (johnsonj) [inline src/route-sync/cloudfoundry/router.go]: addressed

File: src/route-sync/cloudfoundry/router.go (hunk 12 above)
Comment (johnsonj) [inline src/route-sync/cloudfoundry/router.go]: addressed

File: src/route-sync/main.go [hunk 26]
```diff
@@ -0,0 +1,101 @@
+package main
```
Comment (johnsonj) [inline src/route-sync/main.go]: mostly, addressed

File: src/route-sync/cloudfoundry/router.go (hunk 24 above)
Comment (johnsonj) [inline src/route-sync/cloudfoundry/router.go]: addressed, we now return the error from nats

File: src/route-sync/pooler/pooler.go [hunk 27]
```diff
@@ -0,0 +1,86 @@
...
+
+// ByTime returns a Pooler that refreshes every duration
+func ByTime(duration time.Duration, logger lager.Logger) Pooler {
+	return &timeBased{duration: duration, running: false, logger: logger}
+}
+
+func (tb *timeBased) tick(src route.Source, router route.Router) {
+	tcpRoutes, err := src.TCP()
+	if err != nil {
+		tb.logger.Fatal("fetching TCP routes", err)
```
Comment (johnsonj) [inline src/route-sync/pooler/pooler.go]: addressed

File: src/route-sync/pooler/pooler.go [hunk 28]
```diff
@@ -0,0 +1,86 @@
...
+	}
+	httpRoutes, err := src.HTTP()
+	if err != nil {
+		tb.logger.Fatal("fetching HTTP routes", err)
+	}
+	err = router.HTTP(httpRoutes)
+	if err != nil {
+		tb.logger.Fatal("posting HTTP routes", err)
+	}
+	tb.logger.Info("registered routes", lager.Data{
```
Comment (johnsonj) [inline src/route-sync/pooler/pooler.go]: addressed

**Comment by johnsonj** [comment]
Signing off on this PR. Comments addressed.

@mkjelland can you sign off if ready?

Leaving for @alex-slynko / @jhvhs to merge
